examples/gno.land/p/demo/btree/btree.gno

```diff
@@ -0,0 +1,1114 @@
+//   children are copied to ensure they share the correct context, enabling safe and
+//   isolated updates without affecting other trees that might be referencing the original nodes.
+//
+// Example Usage:
+// When a tree performs a write operation (e.g., inserting or deleting a node), it uses
+// its copyOnWriteContext to determine whether it can modify nodes directly or needs to
+// create copies. This mechanism ensures that trees can share nodes efficiently while
+// maintaining data integrity.
+type copyOnWriteContext struct {
+	nodes *FreeNodeList
+}
+
+// Record implements an interface with a single function, Less. Any type that implements
+// RecordIterator allows callers of all of the iteration functions for the BTree
+// to evaluate an element of the tree as it is traversed. The function will receive
+// a stored element from the tree. The function must return either a true or a false value.
+// True indicates that iteration should continue, while false indicates that it should halt.
+type RecordIterator func(i Record) bool
+
+//////////
+//
+// Functions
+//
+//////////
+
+// NewFreeNodeList creates a new free list.
+// size is the maximum size of the returned free list.
+func NewFreeNodeList(size int) *FreeNodeList {
+	return &FreeNodeList{nodes: make([]*node, 0, size)}
+}
+
+func (freeList *FreeNodeList) newNode() (nodeInstance *node) {
+	index := len(freeList.nodes) - 1
+	if index < 0 {
+		return new(node)
+	}
+	nodeInstance = freeList.nodes[index]
+	freeList.nodes[index] = nil
+	freeList.nodes = freeList.nodes[:index]
+
+	return nodeInstance
+}
+
+// freeNode adds the given node to the list, returning true if it was added
+// and false if it was discarded.
+
+func (freeList *FreeNodeList) freeNode(nodeInstance *node) (nodeWasAdded bool) {
+	if len(freeList.nodes) < cap(freeList.nodes) {
+		freeList.nodes = append(freeList.nodes, nodeInstance)
+		nodeWasAdded = true
+	}
+	return
+}
+
+// A default size for the free node list. We might want to run some benchmarks to see if
+// there are any pros or cons to this size versus other sizes. This seems to be a reasonable
+// compromise to reduce GC pressure by reusing nodes where possible, without stacking up too
+// much baggage in a given tree.
+const DefaultFreeNodeListSize = 32
+// New creates a new B-Tree with the given degree. If a degree is not provided,
+// it will default to 16 element nodes. Degree may not be less than 1 (which
+// effectively makes the tree into a binary tree).
+//
+// New(2), for example, will create a 2-3-4 tree (each node contains 1-3 records
+// and 2-4 children).
+func New(params ...int) *BTree {
+        if len(params) == 0 {
+		return NewWithFreeNodeList(16, NewFreeNodeList(DefaultFreeNodeListSize))
+	} else if len(params) == 1 {
+		return NewWithFreeNodeList(params[0], NewFreeNodeList(DefaultFreeNodeListSize))
+	} else {
+		return NewWithFreeNodeList(params[0], NewFreeNodeList(params[1]))
+	}
```

Comment:
This constructor is not explicit and is attempting to use variadic arguments for "overloading". This is generally bad practice in Go.

Can we simply make this work in the case for `New()`, just returning `NewWithFreeNodeList(16, NewFreeNodeList(DefaultFreeNodeListSize))`?

Reply:
This is fair. I am far more accustomed to languages that allow for some style differences from Go that I have hit my head against a few times.

However, since I wrote this, I've learned a thing or two. I'll revise so that this fits a standard Go pattern.

Reply:
So, what I ended up doing is implementing a functional configuration system.

```go
tree := btree.New(WithDegree(10))
```

```go
tree := btree.New(
        WithDegree(12),
        WithFreeNodeList(NewFreeNodeList(100)))
```

Without any functional parameters, the default is a degree of 16, and a new free node list of 32 elements.